File: myLeetCode/Monotone_stack/739. Daily Temperatures.cpp
// 739. Daily Temperatures
// https://leetcode.com/problems/daily-temperatures/

#include <iostream>
#include <vector>
#include <stack>

std::vector<int> dailyTemperatures(std::vector<int>& temperatures) {
    int n = temperatures.size();
    std::vector<int> result(n, 0);
    std::stack<int> st;

    for (int i = 0; i < n; i++) {
        while (!st.empty() && temperatures[i] > temperatures[st.top()]) {
            int index = st.top();
            st.pop();
            result[index] = i - index;
        }
        st.push(i);
    }

    return result;
}

int main() {
    std::vector<int> temperatures = {73, 74, 75, 71, 69, 72, 76, 73};

    std::vector<int> result = dailyTemperatures(temperatures);

    std::cout << "Number of days until a warmer temperature: ";
    for (int i = 0; i < result.size(); i++) {
        std::cout << result[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: C++/代码随想录/Monotone_stack/739. Daily Temperatures.py
def dailyTemperatures(T):
    result = [0] * len(T)
    stack = []  # 存储索引

    for i, temp in enumerate(T):
        while stack and T[stack[-1]] < temp:
            prev_index = stack.pop()
            result[prev_index] = i - prev_index
        stack.append(i)

    return result

# 测试用例
print(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]))  # 应输出 [1, 1, 4, 2, 1, 1, 0, 0]
